### PR TITLE
Parse RDF's and find minima

### PR DESCRIPTION
**Pull request by orioncohen**
## Description
Provide a brief description of the PR's purpose here.

## Todos
- [x] add test data
- [x] functionality to interpolate rdfs
- [x] functionality to find minima
- [x] refine minima finding
- [x] solidify unit testing

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by orioncohen** [comment]
Looks like by rebasing I have some extra commits in the commit history. The first 4 commits in this PR are already incorporated into main.

**Comment by hmacdope** [comment]
General note that you should add a README with any test data to say how you generated it :)

File: solvation_analysis/rdf_parser.py
Comment (hmacdope) [inline solvation_analysis/rdf_parser.py:15]: This looks like a really promising approach. It is also readily expandable to any dimension with in a multidimensional spline basis.

The general case is any function that maps
```
f(rdf[:])-> critical_points[:]
```
I think splines are a really good candidate for this.

File: solvation_analysis/rdf_parser.py
Comment (hmacdope) [inline solvation_analysis/rdf_parser.py:7]: Just a notation thing, the opposite of `floor` is often called `ceil`.

Comment (orioncohen) [reply solvation_analysis/rdf_parser.py:7]: So these aren't actually opposites, though admittedly the naming I used here is a bit confusing. Floor is used to detect the leading edge of the first peak, so its a y-axis point. Cutoff is the end of the fitting region, so its a x-axis point. Very open to alternative naming.

File: solvation_analysis/rdf_parser.py
Comment (orioncohen) [reply solvation_analysis/rdf_parser.py]: The results are looking promising! My plan is to compose the function suggested above from the `interpolate_rdf` and `identify_minima` functions.

**Comment by orioncohen** [comment]
General question here. For the interpolation, much of the testing and feedback process is highly visual. What is the best way to test or at least demonstrate the visual aspect of this code? Should I have the unit tests write plots to a folder that can then be scrutinized? Or should I just save some plots manually and update them periodically?

File: solvation_analysis/rdf_parser.py
Comment (hmacdope) [reply solvation_analysis/rdf_parser.py:7]: That makes sense, I'm happy to leave them as they are.

File: solvation_analysis/rdf_parser.py
Comment (hmacdope) [reply solvation_analysis/rdf_parser.py]: perfecto!

**Comment by hmacdope** [comment]
WRT the plots, I personally would make a diagnostic jupyter notebook with your workflow included that can be easily run on your test data. That way if someone wants to check the values used in the tests, its all set up to be validated easily by eye. The values in the tests can then be reliably retested. Perhaps @richardjgowers or @IAlibay have an opinion also. 

WRT README, just in the `tests/data`  folder is fine.

File: requirements.txt
Comment (IAlibay) [inline requirements.txt:7]: Is this necessary? Ideally we want to match the dependency range that someone using the same MDA release would use.

File: solvation_analysis/tests/data/rdf_non_solvated/rdf_bn_N_vs_fec_F_bins.csv
Comment (IAlibay) [inline solvation_analysis/tests/data/rdf_non_solvated/rdf_bn_N_vs_fec_F_bins.csv:1]: Would npz files be a better format to store this in than csv?

File: solvation_analysis/rdf_parser.py
Comment (IAlibay) [inline solvation_analysis/rdf_parser.py:19]: This error message gets outputted if the function isn't a instance, is this sufficiently cleaer for users to understand what's happening?

Also it's more lines, but I'd suggest avoiding the assert here and going for an if/else with a raise. This will let you set the raise to an error that is more sensible (AssertionError is not what folks will be expecting, maybe a TypeError?)

File: solvation_analysis/tests/conftest.py
Comment (IAlibay) [inline solvation_analysis/tests/conftest.py:104]: The following three are very similar / seem to be using the same code. Can you either a) parametrize the fixture, or b) just create a helper function that just takes in the relevant bins & data references and returns the dictionaries?

File: solvation_analysis/rdf_parser.py
Comment (richardjgowers) [inline solvation_analysis/rdf_parser.py:25]: For a plotting function, I'd rather it accept an "ax" argument which is a Matplotlib axis, then it works on that and returns it. Plt.plot works in mysterious ways I'd rather avoid.

File: solvation_analysis/rdf_parser.py
Comment (richardjgowers) [reply solvation_analysis/rdf_parser.py:19]: I'd sooner just accept anything, put it inside a try/except and complain later. It'd be annoying if I were interpolating differently and this function refused to work for me.

File: solvation_analysis/tests/data/rdf_non_solvated/rdf_bn_N_vs_fec_F_bins.csv
Comment (richardjgowers) [reply solvation_analysis/tests/data/rdf_non_solvated/rdf_bn_N_vs_fec_F_bins.csv:1]: I'd be curious if gzip would be smaller than npz 🙃

**Comment by richardjgowers** [comment]
The tests don't need to visually check each time. Run it once locally visually, check you're happy then encode the minima (eg the correct answer is 4.5) into the test. Only changes to the algorithm(a) should then break this test. This is a little more like a regression test.

File: solvation_analysis/tests/data/rdf_non_solvated/rdf_bn_N_vs_fec_F_bins.csv
Comment (IAlibay) [reply solvation_analysis/tests/data/rdf_non_solvated/rdf_bn_N_vs_fec_F_bins.csv:1]: Bit old, but looks like someone actually went through the trouble: https://stackoverflow.com/a/41425878 (I assume csv + gzip is what you were proposing?)

Comment (hmacdope) [reply solvation_analysis/tests/data/rdf_non_solvated/rdf_bn_N_vs_fec_F_bins.csv:1]: I think npz files are a good idea, as they will always be well supported by numpy, and can be read in directly,

**Comment by hmacdope** [comment]
@orioncohen I would prefer to have something that "works" for finding minima (as you have demonstrated on some test data) and move on to building additional functionality around that than deeply explore the optimal minima finding setup. Once you have working tests for the "easy" and "hard" rdf cases and have addressed #27  please ping! :)

File: solvation_analysis/rdf_parser.py
Comment (hmacdope) [inline solvation_analysis/rdf_parser.py:40]: AFAIK this only identifies a single solvation cutoff, this should defs be extensible to as many as you want, ie an array of minima.  This just means you need to something along the lines of:

```python 
#check that there is a min and max for each point
#this may not be true @orion and you will know if you need to check differently
assert(len(cr_points)%2 == 0)
minima = np.empty(len(cr_points)/2)
for i in range(0,len(cr_points),2):
	#assume max always comes first and that it goes max-min-max-min
	max = cr_points[i]
	min = cr_points[i+1]
	if max <= min:
		#some error path here
		raise ShockinglyHorribleInterpolationException
	else:
		minima[i] = min
#some other stuff
return minima
```
That way you can get a set of solvation critical points rather than just the first one. 
This is by no means the best solution, but I thought i would try and convey the idea in some (pseudo) code.

File: solvation_analysis/tests/conftest.py
Comment (hmacdope) [reply solvation_analysis/tests/conftest.py:104]: Seconded.

File: solvation_analysis/tests/test_rdf_parser.py
Comment (hmacdope) [inline solvation_analysis/tests/test_rdf_parser.py:51]: As you know needs some asserts :)

**Review by hmacdope — CHANGES_REQUESTED** [review]
Good progress so far, see suggestions about finding minima within our min-max min-max... approximation. :)

File: solvation_analysis/tests/conftest.py
Comment (orioncohen) [reply solvation_analysis/tests/conftest.py]: Done

File: solvation_analysis/tests/data/rdf_non_solvated/rdf_bn_N_vs_fec_F_bins.csv
Comment (orioncohen) [reply solvation_analysis/tests/data/rdf_non_solvated/rdf_bn_N_vs_fec_F_bins.csv]: After mistakenly implementing this with .npy files, I now fixed it and all the data is .npz !

File: requirements.txt
Comment (orioncohen) [reply requirements.txt:7]: It's not necessary, should I remove the specific versioning altogether or specify versions to match the MDA release?

Comment (hmacdope) [reply requirements.txt:7]: @IAlibay has the final say here, but I would lean towards matching the MDA 2.0.0-dev requirements.

File: solvation_analysis/tests/data/README.md
Comment (hmacdope) [inline solvation_analysis/tests/data/README.md:28]: Perfect README, and good idea to have an RDF which is a mess that should not register.

File: solvation_analysis/rdf_parser.py
Comment (orioncohen) [reply solvation_analysis/rdf_parser.py:25]: Good call. I will update to use fig, ax.

File: solvation_analysis/tests/data/README.md
Comment (orioncohen) [reply solvation_analysis/tests/data/README.md:28]: I'll build in some failing cases soon, my recent efforts have just been getting test data together and fomatted reasonably

File: requirements.txt
Comment (orioncohen) [reply requirements.txt]: I matched the MDA requirements specified on the MDA Github. Let me know how the new version looks.

File: solvation_analysis/rdf_parser.py
Comment (orioncohen) [reply solvation_analysis/rdf_parser.py]: Got it, I threw all this inside a try catch that should catch anything with a .derivative() method and give a slightly more informative error. Let me know if the new solution looks alright.

File: solvation_analysis/rdf_parser.py
Comment (orioncohen) [reply solvation_analysis/rdf_parser.py:25]: Agreed, I personally loath matplotlibs stateful "plt.plot" interface but I was rushing.

File: solvation_analysis/rdf_parser.py
Comment (orioncohen) [reply solvation_analysis/rdf_parser.py:41]: I agree in general but I think I need to improve the minima identifying method before generalizing to further solvation cutoffs beyond the first. The min-max min-max approach isn't good enough to weed out very slight split peaks and small divets in the fit. I could try applying a stronger smoothing factor to the interpolation or throwing in an additional peak fitting step, but as you suggested, I think it's best to get something that works and start developing the analysis framework. 

For now, I think I should throw together a heuristic for the first solvation cutoff and move on. I can circle back to improving the fitting algorithm later. I renamed issue #19 to reflect this, I think it is closely related.

File: solvation_analysis/rdf_parser.py
Comment (hmacdope) [inline solvation_analysis/rdf_parser.py:115]: Docs say that `cutoff_region` is a tuple with an example of length 2 given, here we access the 5th element with `cutoff[4]`. Am I missing something here?

File: solvation_analysis/rdf_parser.py
Comment (hmacdope) [reply solvation_analysis/rdf_parser.py:122]: Good decision I think to move forward while documenting the limitations. As you say I think we want to develop the analysis framework rather than focusing on the perfect minima finding heuristic.

File: requirements.txt
Comment (hmacdope) [reply requirements.txt]: LGTM but as before @IAlibay  is the expert on this area of MDA.

File: solvation_analysis/rdf_parser.py
Comment (hmacdope) [inline solvation_analysis/rdf_parser.py:51]: I would perhaps change this to:
```python
try:
 isinstance(scipy.interpolate.UnivariateSpline)
except:
	ValueError("blah")
cr_pts = f.derivative.roots() #blah blah
```
This is mainly because I don't like having the majority of work done inside a `try/except` block, but this is a personal thing. If you like it the way it is there really isn't anything wrong with it.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Just some very minor things.

File: solvation_analysis/rdf_parser.py
Comment (orioncohen) [reply solvation_analysis/rdf_parser.py]: Nope! That was a typo that I would have caught had I been doing TDD better :)

Fixed it and added testing.

**Comment by orioncohen** [comment]
All of my unit tests are passing locally but many are failing in CI. Specifically, all of the tests that use the RDF data are failing. When the tests try to get the `bins` and `rdf` arrays from the dictionary that stores filenames, there is a key error. For example, the code below throws a key error in CI:
```
def test_identify_solvation_cutoff(rdf_tag, cutoff, rdf_bins_and_data_easy):
    bins, rdf = rdf_bins_and_data_easy[rdf_tag]
```
The `rdf_bins_and_data_easy` is a fixture that is generated from a dictionary of filenames in `datafiles.py`. The fixture itself seems unlikely to fail, so what I think is going on here is that the code I wrote to retrieve the rdf data files is not searching in the right place. My code uses: 

```
data_dir = Path("data")
easy_rdf_dir = data_dir / "rdf_vs_li_easy"

easy_rdf_bins = {
    generate_tag(rdf_path): resource_filename(__name__, str(rdf_path))
    for rdf_path in easy_rdf_dir.glob("*bins.npz")
}
```
where
```
def generate_tag(filename):
    # generates a ad-hoc tag for the rdf data from the file name
    tag_list = filename.stem.split("_")
    rdf_tag = f"{tag_list[1]}_{tag_list[2]}"
    return rdf_tag
```
I believe the problem is that when I try to call `easy_rdf_dir.glob("*bins.npz")`, I am not actually iterating through the right directory. If the dictionary ends up empty, that would explain the key errors.

I will keep tinkering around but I don't quite understand how CI manages file access. How can I make sure that I am iterating through the right directory? How does CI set up the environment?

**Comment by orioncohen** [comment]
Testing is now in a pretty good place I think. The cutoff identification algorithm will absolutely need more work later but it is at least functioning. I think this is about ready to merge and move onto the analysis sections.

**Comment by lilyminium** [comment]
@orioncohen I wouldn't rely on finding `"data"`. I think that's always checked relative to the working directory, which could be literally anywhere (e.g. `pytest my/very/long/path/to-/solvation-analysis/tests`). Try making it relative to `__file__`, which is absolute. (e.g. `data_dir = Path(__file__) / "data"`)

**Comment by orioncohen** [comment]
Thanks @lilyminium! Everything is passing now.

File: solvation_analysis/rdf_parser.py
Comment (tylerjereddy) [inline solvation_analysis/rdf_parser.py:21]: typo "probablility"

File: solvation_analysis/rdf_parser.py
Comment (tylerjereddy) [inline solvation_analysis/rdf_parser.py:100]: `fail` -> `fails` maybe

File: solvation_analysis/rdf_parser.py
Comment (tylerjereddy) [inline solvation_analysis/rdf_parser.py:108]: `value` -> `values` maybe

**Comment by hmacdope** [comment]
Awesome work @orioncohen! Just address @tylerjereddy's last comments and we are good to rock.

**Comment by orioncohen** [comment]
Typos are fixed! This should be ready to merge.

Thanks for the proofreading @tylerjereddy :)

**Review by hmacdope — APPROVED** [review]
Awesome work! This has come along very nicely. @orioncohen you will need to convert from draft to active PR for this to be merged, but LGTM. :)

**Comment by hmacdope** [comment]
Also wait for CI to be green.

**Comment by hmacdope** [comment]
@orioncohen you can convert from draft and merge when ready.

**Comment by orioncohen** [comment]
For some reason the CI checks for windows and macOS seem to be stalling. They fail due to timeout after 6 hours running. Any ideas on this? Should I ignore it and merge? The linux tests are passing without issue.

File: requirements.txt
Comment (IAlibay) [inline requirements.txt:3]: Your test environment is still pinned to 1.1.1, setting to 2.0.0 here doesn't seem necessary?

**Review by IAlibay — COMMENTED** [review]
The fact that MacOS and Windows are failing really need investigating. I'm not sure what's the cause but I wouldn't just merge as-is.

It seems to be stalling in the tests and I can't seem to be able to get the full logs.

Off the top of my head the likely culprits here are:
1. Some issue with the MPL backend
2. Some issue reading in npz files

What I suggest you do is:
1. Disable all the new tests see if CI completes
2. Slow re-enable different tests until it stalls.

**Comment by hmacdope** [comment]
@orioncohen I have re-trigged the jobs to see if anything changes.

**Comment by orioncohen** [comment]
Thanks for the tip @IAlibay.

Looks like the issue was with plt.show(). This is only useful locally anyway so I am going to comment out plt.show() and make a note that it should be uncommented for local testing.

**Comment by IAlibay** [comment]
> Thanks for the tip @IAlibay.
> 
> Looks like the issue was with plt.show(). This is only useful locally anyway so I am going to comment out plt.show() and make a note that it should be uncommented for local testing.

@orioncohen 

If it's untestable and isn't offering a very high level of complexity (from what I can see both that function and the nglview one are just simple mpl/nglview calls), then I suggest just removing them from your codebase. In my opinion you're much better off having good documentation that shows a worked example of how you can generate that plot rather than have the extra code complexity of "uncomment this if X, maybe".

(Similarly nglview is a reasonably large dependency, I like it, but it's probably not worth enforcing users need to download it).

**Pull request merged**
